File: python/gradbench/gradbench/tools/jax/lse.py
import jax
import jax.numpy as jnp
import numpy as np
from gradbench import wrap
from jax import grad, jit

jax.config.update("jax_enable_x64", True)


def logsumexp(x):
    x_max = jnp.max(x)
    return jnp.log(jnp.sum(jnp.exp(x - x_max))) + x_max


def prepare_input(input):
    return np.array(input["x"], dtype=np.float64)


@wrap.multiple_runs(pre=prepare_input, post=float)
@jit
def primal(input):
    return logsumexp(input)


@wrap.multiple_runs(pre=prepare_input, post=lambda x: x.tolist())
@jit
def gradient(input):
    return grad(logsumexp)(input)

File: js/tensorflow/lse.ts
import type { Runs } from "@gradbench/common/protocol.ts";
import { multipleRuns } from "@gradbench/common/util.ts";
import * as tf from "@tensorflow/tfjs";

const f = (x: tf.Tensor): tf.Tensor => {
  const a = x.max();
  return a.add(x.sub(a).exp().sum().log());
};

interface Input extends Runs {
  x: number[];
}

export const primal = multipleRuns(({ x }: Input) => {
  return () => f(tf.tensor(x)).arraySync();
});

export const gradient = multipleRuns(({ x }: Input) => {
  const g = tf.grad(f);
  return () => g(tf.tensor(x)).arraySync();
});
